SANC-48-When-the-user-completes-registration-they-should-be-able-to-input-their-name

jason-duong4509 opened the pull request:
- Completed SANC-48
- Added a loading screen while the Google Maps API loads (feedback from SANC-39). Found in \src\app\_components\agencycomponents\agency-interactive-area.tsx

Attached is a video demonstrating the login process, starting when the user has a name (directs them to /dashboard) and then when the user has no name (directs them to /fill-out-name).

In the video, /dashboard was a copy-paste of the agency-interative-area.tsx file. It was only added since the endpoint does not yet exist. It is not included in the final branch (/dashboard remains non-existent)

https://github.com/user-attachments/assets/3abf4830-181c-4425-a4d6-155bf999d5c8

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

* **New Features**
  * Added a "Fill Out Name" page for users to enter and save their profile name and be redirected after saving.
  * New server endpoint to change a user's name and support the post-save redirect flow.

* **Improvements**
  * Users without a name are routed to the name entry page before dashboard access.
  * Loading now shows a styled, centered loading card instead of an empty placeholder.

* **Validation**
  * Name input enforces format rules to prevent invalid entries.

<sub>✏️ Tip: You can customize this high-level summary in your review settings.</sub>
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

jason-duong4509 committed:
SANC-48: Completed Ticket

- Completed SANC-48
- Added a loading screen while the Google Maps API loads (feedback from SANC-39). Found in \src\app\_components\agencycomponents\agency-interactive-area.tsx

jason-duong4509 committed:
SANC-48: Minor Quick Fixes

- Changed the function name in /src/app/fill-out-name from LoginPage to FillOutNamePage because that was missed earlier
- Added a comment in /src/app/fill-out-name because that was missed earlier
- Changed an HTTP error code from 500 to 400 because of the coderabbitai suggestion

jason-duong4509 commented on src/app/fill-out-name/page.tsx:
```diff
@@ -0,0 +1,75 @@
+      //User inputted name is not proper
+      notify.error("Invalid Name");
+      setLoading(false);
+    } else {
```
3 character names do work. Resolving conversation as a result

jason-duong4509 commented on src/server/api/routers/organizations.ts:
```diff
@@ -224,4 +226,34 @@ export const organizationRouter = createTRPCRouter({
+
+  changeName: protectedProcedure
+    .input(z.object({ newName: z.string() }))
+    .mutation(async ({ input, ctx }) => {
+      const { newName } = input; //Get the passed variables
+
+      //Alphanumeric regex of length >=3. Allows for exactly 1 space between characters
+      const regex = /^(?!.*\s{2})[A-Za-z0-9][A-Za-z0-9\s]*[A-Za-z0-9][A-Za-z0-9\s]*[A-Za-z0-9]$/;
+      if (!regex.test(newName)) {
+        //User inputted name is not proper
+        throw new TRPCError({
+          code: "INTERNAL_SERVER_ERROR",
+          message: "Invalid Name",
+        });
+      }
+
+      const listOfUpdatedUsers = await ctx.db
+        .update(user)
+        .set({ name: newName })
+        .where(eq(user.id, ctx.session.user.id))
+        .returning(); //Make DB update command to change the user's name
+
+      if (listOfUpdatedUsers.length === 0) {
+        //Drizzle updated no users (something went wrong)
+        throw new TRPCError({
+          code: "INTERNAL_SERVER_ERROR",
+          message: "Failed to update user's name",
+        });
+      }
+    }),
```
1. done
2. likely not needed 
3. does work for 3 characters

burtonjong commented on src/app/fill-out-name/page.tsx:
```diff
@@ -0,0 +1,75 @@
+  const handleSubmit = async (values: typeof form.values) => {
+    setLoading(true);
+    //Alphanumeric regex of length >=3. Allows for exactly 1 space between characters
+    const regex = /^(?!.*\s{2})[A-Za-z0-9][A-Za-z0-9\s]*[A-Za-z0-9][A-Za-z0-9\s]*[A-Za-z0-9]$/;
```
You should be able to add this validation to `validation.ts` and then move the actual validation checking to the useForm() call (because zod will handle it).

jason-duong4509 replied:
Done!

burtonjong commented on src/app/fill-out-name/page.tsx:
```diff
@@ -0,0 +1,75 @@
+    },
+  });
+
+  const changeNameMutation = api.organization.changeName.useMutation({
```
If I'm not mistaken, you should be able to extract a loading state from this mutation instead of making your own with useState().

jason-duong4509 replied:
Can you tell me more about this? I am not following

burtonjong replied:
I believe you should be able to do `changeNameMutation.isPending` instead of having a loading useState.

jason-duong4509 replied:
Sounds good, made that change

burtonjong commented on src/server/api/routers/organizations.ts:
```diff
@@ -224,4 +226,34 @@ export const organizationRouter = createTRPCRouter({
+        });
+      }
+
+      const listOfUpdatedUsers = await ctx.db
```
list of updated users? If you only update the user where `user.id` === `ctx.session.user.id` shouldn't there only be one anyway?

jason-duong4509 replied:
The naming was just for me to remember that an `UPDATE` with `RETURNING` gives a list of all affected users. In this case, it's always a list of size 1 or 0 (if the DB could not find a user to update), like you mentioned

Later in that code snippet, `listOfUpdatedUsers` is checked to see whether the DB was able to update a user or not. That is the only purpose of that variable

burtonjong commented on src/lib/auth.ts:
```diff
@@ -11,6 +12,19 @@ export const auth = betterAuth({
+          //Account name not set
+          return { newLink: "/fill-out-name" };
+        }
+        //TODO V: change later when we know where to send the user
```
is this `return { newLink : "/dashboard" };` just how nextauth suggests that you redirect the user on a serverside call like this?

jason-duong4509 replied:
No. Assuming you mean betterauth (unless they're the same thing), betterauth suggests using a throw to redirect, `throw ctx.redirect("/sign-up/name");`. However, since the login uses a fetch call, that throw will cause the server to redirect, not the browser (`result.data` ends up containing the code of `/fill-out-name` or `/dashboard` I believe). The browser does not redirect when the backend uses throw and fetch is used

From what I found looking into it, the typical way of doing it is to have the after hook return success on a good login and have the frontend redirect itself (like what was previously in the code). In this case, though, the frontend won't know whether to navigate to `/dashboard` or `/fill-out-name` since the backend determines that. I chose to have the backend return a JSON containing the endpoint to redirect to

Upon looking at it some more, I think the after hook can be eliminated entirely and the frontend can take care of where to redirect the user upon successful login. Both ways achieve the same result; I previously wasn't able to figure out a way to get the frontend to know where to redirect. I can change the code and make the frontend redirect properly if desired

burtonjong replied:
@jason-duong4509 I think I would prefer the frontend redirects, not the server. Move that functiionality out and then this looks good

jason-duong4509 replied:
Done!

promatty commented:
@burtonjong

jason-duong4509 committed:
SANC-48: Made Fixes to Address Feedback

- Moved front-end redirect logic from the back-end to the front-end
- Replaced the loading useState
- Moved the name regex to validation.ts
- Simplified form validation in \fill-out-name\page.tsx to streamline code

jason-duong4509 committed:
Merge remote-tracking branch 'origin/main' into SANC-48-When-the-user-completes-registration-they-should-be-able-to-input-their-name

# Conflicts:
#	src/app/_components/agencycomponents/agency-interactive-area.tsx

jason-duong4509 committed:
Merge remote-tracking branch 'origin/main' into SANC-48-When-the-user-completes-registration-they-should-be-able-to-input-their-name

# Conflicts:
#	src/app/login/page.tsx
#	src/server/api/routers/organizations.ts

jason-duong4509 committed:
SANC-48: Made Fixes due to Merge Conflict

- Changed the way a user is redirected from /login to /fill-out-name to accommodate changes made to main
- Changed the way a user is redirected from /fill-out-name to accommodate changes made to main

jason-duong4509 commented:
@burtonjong

burtonjong commented on src/app/fill-out-name/page.tsx:
```diff
@@ -20,11 +20,19 @@ export default function FillOutNamePage() {
     },
   });
 
+  const { mutate } = api.organization.redirectToDashboard.useMutation({
```
calling this function `mutate()` is misleading. A more descriptive name is better

jason-duong4509 replied:
Done!

burtonjong commented on src/server/api/routers/organizations.ts:
```diff
@@ -31,50 +31,54 @@ export const organizationRouter = createTRPCRouter({
+          if (!activeOrgId) {
+            throw new TRPCError({
+              code: "NOT_FOUND",
+              message: "No active organization found for this agency user",
```
Maybe add to this at the end: please contact an administator for help... something like that

jason-duong4509 replied:
Done!

burtonjong commented on src/server/api/routers/organizations.ts:
```diff
@@ -31,50 +31,54 @@ export const organizationRouter = createTRPCRouter({
+          if (!organization.slug) {
+            throw new TRPCError({
+              code: "INTERNAL_SERVER_ERROR",
+              message: "Organization is missing required slug configuration",
```
Same thing here as above

jason-duong4509 replied:
Done!

jason-duong4509 committed:
SANC-48: Made Changes to Address Feedback

- Changed a function name for better clarity
- Added additional text to error messages to provide more information

jason-duong4509 committed:
SANC-48: Hotfix

- Changed how the back-end validated user name input by relying on an
  imported regex rather than a hardcoded one in-file. Missed this issue
  in previous commits

jason-duong4509 commented on src/server/api/routers/organizations.ts:
```diff
@@ -29,50 +31,56 @@ export const organizationRouter = createTRPCRouter({
-            message: "No active organization found for this agency user",
+    if (ctx.session.user.name === "") {
+      redirectUrl = "/fill-out-name";
+    } else {
```
This does not actually happen. Resolving comment as a result

burtonjong commented on src/server/api/routers/organizations.ts:
```diff
@@ -296,4 +304,32 @@ export const organizationRouter = createTRPCRouter({
+        });
+      }
+
+      const listOfUpdatedUsers = await ctx.db
+        .update(user)
+        .set({ name: newName })
+        .where(eq(user.id, ctx.session.user.id))
+        .returning(); //Make DB update command to change the user's name
+
+      if (listOfUpdatedUsers.length === 0) {
+        //Drizzle updated no users (something went wrong)
+        throw new TRPCError({
+          code: "INTERNAL_SERVER_ERROR",
+          message: "Failed to update user's name",
+        });
+      }
+    }),
```
@jason-duong4509 This is probably going to end up being a problem. Updating the db user name and not updating the session user name will cause an inconsistency.

On line 20 we already get the user. We can just use that user instead of ctx.session.user.name, unless you thinbk otherwise

jason-duong4509 replied:
Done!

jason-duong4509 committed:
SANC-48: Made Changes to Address Feedback

- Changed how the redirectToDashboard endpoint checked whether a user's
  name was empty by using the user's name from the DB rather than the
  current session

jason-duong4509 committed:
Merge remote-tracking branch 'origin/main' into SANC-48-When-the-user-completes-registration-they-should-be-able-to-input-their-name

# Conflicts:
#	src/types/validation.ts
